keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+
+2. That Kube-proxy enables connection draining for terminating Nodes. This is to
+  be done by Kube-proxy by inspecting a field on the Node object which indicates
+  that the "node is terminating/deleting" and when seen, starts failing the HC.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+We need to agree on this field. `deletionTimestamp` is the perfect indicator for
+this, but....the Node object does not have a finalizer, so that wouldn't help.
+@thockin advocates for `.spec.unschedulable`, but I am inclined to say "that
+doesn't indicate if a Node is going away". `unschedulable` just says that no
+workloads can be assigned to the Node anymore. It is true that it's usually
+followed up by a drain...but I am not sure it's good to implement something
+which can easily kill ingress connectivity to a Node / cluster by simply doing
+`kubectl cordon node ...` (It would kill ingress connectivity to the entire
+cluster if the ingress controller is running on the node which is being cordoned)
+
+Maybe we need a KEP for a finalizer on the Node object (*troll face)
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+Figure out which other cloud-providers configure the HC to target the service
+proxy for the HC
+<<[/UNRESOLVED]>>
+```
+
+## Motivation
+
+The motivation for each change is:
+
+1. Not including the `healthz` state can cause Kube-proxy to indicate to a load
+  balancer that it should send traffic to a Node simply because the endpoint is
+  scheduled there, even though Kube-proxy might not be healthy and successfully
+  managed to write the rules required for actually being able to forward traffic
+  to the endpoint.
+
+2. Workloads on terminating Nodes would benefit from connection draining as to
+  have their connections gracefully terminated. Workloads running on terminating
+  nodes get drained simultaneously with Kube-proxy today, this means that
+  there's no guarantee that connection draining will be enabled for them
+  currently. That might be the case if Kube-proxy is terminated before the
+  endpoint, but there's no guarantee of that. If Kube-proxy reacts to a "node is
+  about to go away" signal, then Kube-proxy can start prematurely failing the HC
+  and enable connection draining for these workloads.
+
+### Goals
+
+- Have kube-proxy accurately reflect its `healthz` state in its response to LB HCs
+  for class B) services.
+
+- Connection drain terminating Nodes
+
+### Non-Goals
+
+- Aligning cloud provider HCs for class A) services. Cloud providers like
+  Azure/AWS do not configure their HCs to point to the service proxy. Instead
+  they connect to the `NodePort` defined for the service. As to have them
+  benefit from the proposals of this KEP, they would need to change their
+  implementation.
+
+## Proposal
+
+#### Risk
+
+The risk with each proposal is:
+
+1. None. Class B) type services which benefitted from being scheduled on nodes
```

Comment:
I'd say there is risk that the proxy healthz could be flapping or otherwise buggy and this would have disproportionate impact on `eTP: Local` services.  Mitigation here needs to be testing and metrics.

Reply:
on network partitions from the control plane, what will be the behavior on kube-proxy healtcheck? does it change?

if previously kube-proxy healthcheck was returning ok despite it was disconnected, and now changes to return nook, this will mean that the LB will stop to forward traffic despite the pods on the node are able to server ... this was a real thing

Reply:
> I'd say there is risk that the proxy healthz could be flapping or otherwise buggy and this would have disproportionate impact on eTP: Local services. 

This is what find a bit difficult with "risks" on KEPs. I mean: sure if there's a bug with the implementation / or if there are disasters: issues will happen - are we supposed to call them out on the KEP? I thought "risks" means: "assuming the change is implemented correctly and normal operations, these are the risks given these unknown factors", ex [on KEP 1 where we just can't tell what ALL cloud-providers are doing](https://github.com/alexanderConstantinescu/kubernetes-enhancements/blob/50d6306d442aa7c670edaaf47e69e0517f5e02dd/keps/sig-network/3458-remove-transient-node-predicates-from-service-controller/README.md?plain=1#L116-L123)   

> if previously kube-proxy healthcheck was returning ok despite it was disconnected, and now changes to return nook, this will mean that the LB will stop to forward traffic despite the pods on the node are able to server

Yes, with this change we would start returning "not OK" to the LB even if the pod might be running fine, because kube-proxy is not running fine. But I mean: it's just pure luck if any pod benefitted from this before.

Reply:
Risks are "plausible things that could happen which are different after this KEP".  Bugs are plausible.  But it's not just bugs in this implementation, there could be bugs in the rest of kube-proxy which cause it to indicate bad healthz, and this KEP changes eTP: Local to pay attention to that.

So there IS a risk that far-away bugs could hurt services.  

Them we can decide if this is a major risk warranting design changes or not.